onadata/libs/permissions.py

```diff
@@ -300,47 +347,57 @@ class OwnerRole(Role):
 # Memoize a class to permissions dict.
 for role in ROLES.values():
     role.class_to_permissions = defaultdict(list)
-    [role.class_to_permissions[k].append(p) for p, k in role.permissions]
+    for p, k in role.permissions:
+        role.class_to_permissions[k].append(p)
 
 
 def is_organization(obj):
-    """Some OrganizationProfiles have a pointer to the UserProfile, but no
+    """
+    Some OrganizationProfiles have a pointer to the UserProfile, but no
     UserProfiles do. Check for that first since it avoids a database hit.
     """
     try:
-        hasattr(obj, 'userprofile_ptr') or obj.organizationprofile
-        return True
+        return (hasattr(obj, 'userprofile_ptr') or
+                obj.organizationprofile is not None)
     except OrganizationProfile.DoesNotExist:
         return False
 
 
 def get_role(permissions, obj):
-    for role in reversed(ROLES_ORDERED):
-        if role.has_role(permissions, obj):
-            return role.name
+    """
+    Return the user role for the given obj permissions.
+    """
+    for _role in reversed(ROLES_ORDERED):
+        if _role.has_role(permissions, obj):
```

Comment:
On another note, can we avoid this `O(len(ROLES_ORDERED))` loop? I imagine this is affecting performance. If the possible values of `permissions` and `obj` are closed (like we can enumerate them) then we should be able to construct a map once when loading this namespace for the first time, and get constant time access.

Hmm, but we can't do this because `obj` can be anything?

Reply:
Oh wait, but we're looking at the `type` of object to figure this out, is that a closed set?